### PR TITLE
Demo for callback-group-level executor concept.

### PR DESCRIPTION
**Pull request by ralph-lange**
Moved PR from ros2/demos to this repository as discussed in https://github.com/ros2/demos/pull/485.

### CODE REVIEW

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+void PingNode::print_statistics()
+{
+  size_t ping_count = rtt_data_.size();
+
+  size_t high_pong_count = 0;
+  size_t low_pong_count = 0;
+  rclcpp::Duration high_rtt_sum(0, 0);
+  rclcpp::Duration low_rtt_sum(0, 0);
+  for (const auto entry : rtt_data_) {
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp:73]:
```suggestion
  for (const auto & entry : rtt_data_) {
```

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+{
+  rtt_data_[msg->data].high_received_ = now();
+}
+
+void PingNode::low_pong_received(const std_msgs::msg::Int32::SharedPtr msg)
+{
+  rtt_data_[msg->data].low_received_ = now();
+}
+
+void PingNode::print_statistics()
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp:65]:
```suggestion
void PingNode::print_statistics() const
```

File: rclcpp/executors/cbg_executor/include/examples_rclcpp_cbg_executor/ping_node.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+};
+
+class PingNode : public rclcpp::Node
+{
+public:
+  PingNode();
+
+  virtual ~PingNode() = default;
+
+  void print_statistics();
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/include/examples_rclcpp_cbg_executor/ping_node.hpp:45]:
```suggestion
  void print_statistics() const;
```

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+namespace examples_rclcpp_cbg_executor
+{
+
+PingNode::PingNode()
+: rclcpp::Node("ping_node")
+{
+  using std::placeholders::_1;
+  using std_msgs::msg::Int32;
+
+  declare_parameter<double>("ping_period", 0.01);
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp:33]:
```suggestion
  this->declare_parameter<double>("ping_period", 0.01);
```

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+// limitations under the License.
+
+#include "examples_rclcpp_cbg_executor/ping_node.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <memory>
+
+#include <examples_rclcpp_cbg_executor/utilities.hpp>
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/ping_node.cpp:22]: to stay consistent with the pong node
```suggestion
#include "./utilities.hpp"
```

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/pong_node.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    auto endTime = get_current_thread_time() + duration;
+    int x = 0;
+    bool doAgain = true;
+    while (doAgain) {
+      while (x != std::rand() && x % 1000 != 0) {
+        x++;
+      }
+      doAgain = (get_current_thread_time() < endTime);
+    }
+  }
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/pong_node.cpp:87]:
```suggestion
    auto end_time = get_current_thread_time() + duration;
    int x = 0;
    bool do_again = true;
    while (do_again) {
      while (x != std::rand() && x % 1000 != 0) {
        x++;
      }
      do_again = (get_current_thread_time() < end_time);
    }
  }
```

File: rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/pong_node.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+// limitations under the License.
+
+#include "examples_rclcpp_cbg_executor/pong_node.hpp"
+
+#include <cassert>
+
+#include <chrono>
+#include <memory>
+
+#include "examples_rclcpp_cbg_executor/utilities.hpp"
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/pong_node.cpp:22]:
```suggestion
#include "./utilities.hpp"
```

File: rclcpp/executors/cbg_executor/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+find_package(ament_cmake REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(std_msgs REQUIRED)
+
+include_directories(include src)
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/CMakeLists.txt:17]:
```suggestion
include_directories(include)
```

File: rclcpp/executors/cbg_executor/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+ament_target_dependencies(pong_node_only rclcpp std_msgs)
+
+add_executable(ping_and_pong_node src/main.cpp src/examples_rclcpp_cbg_executor/ping_node.cpp src/examples_rclcpp_cbg_executor/pong_node.cpp)
+target_compile_definitions(ping_and_pong_node PRIVATE ADD_PING_NODE=1 ADD_PONG_NODE=1)
+ament_target_dependencies(ping_and_pong_node rclcpp std_msgs)
+
+install(TARGETS ping_node_only pong_node_only ping_and_pong_node
+  DESTINATION lib/${PROJECT_NAME}
+)
+
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/CMakeLists.txt:34]: if you've got a public API, you ought to install it.
```suggestion
install(
  DIRECTORY include/
  DESTINATION include
)
```

File: rclcpp/executors/cbg_executor/package.xml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
+  <name>examples_rclcpp_cbg_executor</name>
+  <version>0.0.1</version>
```
Comment (Karsten1987) [inline rclcpp/executors/cbg_executor/package.xml:5]: I am actually not completely sure about this, but I believe all packages within the same repo have to have the same version number.
```suggestion
  <version>0.10.2</version>
```

**Review by Karsten1987 — COMMENTED** [review]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13661/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8540/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11376/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13726/)

File: rclcpp/executors/cbg_executor/package.xml (hunk 10 above)
Comment (clalancette) [reply rclcpp/executors/cbg_executor/package.xml:5]: Yes, that's correct, they all have to have the same version number.

**Comment by ralph-lange** [comment]
Hi @Karsten1987, I fixed the warnings caused by the line lengths in the CMakeLists.txt. Could you please re-trigger the job on ci.ros2.org?

**Comment by ralph-lange** [comment]
Converted it to a draft until macOS support is implemented.

**Comment by ralph-lange** [comment]
Hi @Karsten1987, I tried to add support for macOS - without having any practical C++ experience on this platform. Could you please test and my implementation?

**Comment by Karsten1987** [comment]
@ralph-lange the CPU affinity indeed required the threads to the suspended while being configured. The condition variable seemed to have work according to the following output:
```
 ➭ ros2 run examples_rclcpp_cbg_executor ping_pong
1614233204.026369 [71]   12615970: using network interface en0 (udp/10.0.0.222) selected arbitrarily from: en0, vnic0, vnic1, vboxnet0, en8
[INFO] [1614233204.055699564] [pong_node]: Running experiment from now on for 10 s ...
[INFO] [1614233214.057528441] [ping_node]: High prio path: Sent 988 pings, received 985 pongs.
[INFO] [1614233214.057547680] [ping_node]: High prio path: Average RTT is 27.9 ms.
[INFO] [1614233214.057557175] [ping_node]: Low prio path: Sent 988 pings, received 0 pongs.
[INFO] [1614233214.057563656] [pong_node]: High priority executor thread ran for 9987 ms.
[INFO] [1614233214.057569745] [pong_node]: Low priority executor thread ran for 0 ms.
```

I could do this without requiring root rights, so maybe that'll also reduce the need for it on Linux.

I further went ahead and refactored the nodes from being compile dependent on the compiler options (`ADD_PING`, `ADD_PONG`, ...) into three individual files. I hope you don't mind that I renamed the executables accordingly.

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13732/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8601/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11442/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13805/)

**Comment by ralph-lange** [comment]
Great that you got the core pinning for macOS working. However, we have to use separate mutexes for each executor instance. Otherwise, with the current lock mechanism, only one of the two executor instances will eventually run. I'll fix this in the evening.

**Comment by ralph-lange** [comment]
@Karsten1987, could you please re-trigger the job on ci.ros2.org?

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13898/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8749/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11607/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13989/)

**Review by Karsten1987 — APPROVED** [review]
this looks good to me with green CI.

However, @clalancette if you could give it a pair of eyes as well. I've ended up contributing to the PR quite a bit.

**Review by clalancette — APPROVED** [review]
Seems reasonable enough to me.  @Karsten1987 I'll let you merge.

**Pull request merged**